src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.Unix/OpenSslX509ChainProcessor.cs

```diff
@@ -352,69 +370,243 @@ internal void CommitToChain()
                     {
-                        refErrors.ClearError(NoCrl);
+                        refErrors.ClearRevoked();
                     }
-                    else if (ocspStatus != NoCrl)
+
+                    if (refErrors.IsRevoked())
                     {
-                        refErrors.ClearError(NoCrl);
-                        refErrors.Add(ocspStatus);
+                        refErrors.ClearRevocationUnknown();
+                        encounteredRevocation = true;
+                        continue;
+                    }
+
+                    if (refErrors.HasRevocationUnknown())
+                    {
+                        refErrors.ClearRevocationUnknown();
+
+                        if (i == 0)
+                        {
+                            statusCode = CheckOcsp(0, _leafHandle, revocationMode);
+                        }
+                        else
+                        {
+                            IntPtr certPtr = Interop.Crypto.GetX509StackField(chainStack, i);
+
+                            using (SafeX509Handle certHandle = Interop.Crypto.X509UpRef(certPtr))
+                            {
+                                statusCode = CheckOcsp(i, certHandle, revocationMode);
+                            }
+                        }
+
+                        if (statusCode != Interop.Crypto.X509VerifyStatusCode.X509_V_OK)
+                        {
+                            refErrors.Add(statusCode);
+                            encounteredRevocation = refErrors.IsRevoked();
+                        }
                     }
                 }
             }
+        }
+
+        private WorkingChain BuildWorkingChain()
+        {
+            if (_workingChain != null)
+            {
+                return _workingChain;
+            }
+
+            WorkingChain workingChain = new WorkingChain();
+            WorkingChain extraDispose = null;
+
+            Interop.Crypto.X509StoreCtxReset(_storeCtx);
+            Interop.Crypto.X509StoreVerifyCallback workingCallback = workingChain.VerifyCallback;
+            Interop.Crypto.X509StoreCtxSetVerifyCallback(_storeCtx, workingCallback);
+
+            bool verify = Interop.Crypto.X509VerifyCert(_storeCtx);
+
+            if (workingChain.AbortedForSignatureError)
+            {
+                Debug.Assert(!verify, "verify should have returned false for signature error");
+                CloneChainForSignatureErrors();
+
+                // Reset to a WorkingChain that won't fail.
+                extraDispose = workingChain;
+                workingChain = new WorkingChain(abortOnSignatureError: false);
+                workingCallback = workingChain.VerifyCallback;
+                Interop.Crypto.X509StoreCtxSetVerifyCallback(_storeCtx, workingCallback);
+
+                verify = Interop.Crypto.X509VerifyCert(_storeCtx);
+            }
+
+            GC.KeepAlive(workingCallback);
```

Comment:
Why is this necessary?  We're referencing workingChain below, and workingCallback is the same as workingChain.VerifyCallback, right?

Reply:
`workingChain.VerifyCallback` is a method(group), `workingCallback` is an instance-bound delegate that was passed as a native callback.

Reply:
Ah, I thought it was a field / field-backed-property.  Nevermind.

Reply:
Added a comment anyways :)